crates/net/eth-wire/src/p2pstream.rs

```diff
@@ -606,32 +565,40 @@ where
         // all messages sent in this stream are subprotocol messages, so we need to switch the
         // message id based on the offset
         compressed[0] = item[0] + MAX_RESERVED_MESSAGE_ID + 1;
-        this.outgoing_messages.push_back(compressed.freeze());
+        this.inner.start_send_unpin(compressed.freeze()).map_err(P2PStreamError::from)?;
 
         Ok(())
     }
 
     /// Returns `Poll::Ready(Ok(()))` when no buffered items remain.
     fn poll_flush(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
         let mut this = self.project();
-        let poll_res = loop {
-            match this.inner.as_mut().poll_ready(cx) {
-                Poll::Pending => break Poll::Pending,
-                Poll::Ready(Err(err)) => break Poll::Ready(Err(err.into())),
+
+        // Check if we need to send a disconnect message or a ping message, if so, send them
+        match (this.pending_disconnect.take(), this.pending_ping.take()) {
+            (Some(disconnect), _ping) => match this.inner.poll_ready_unpin(cx) {
                 Poll::Ready(Ok(())) => {
-                    let Some(message) = this.outgoing_messages.pop_front() else {
-                        break Poll::Ready(Ok(()))
-                    };
-                    if let Err(err) = this.inner.as_mut().start_send(message) {
-                        break Poll::Ready(Err(err.into()))
-                    }
+                    this.inner.start_send_unpin(disconnect)?;
```

Comment:
I see, so disconnect has precedence over ping, makes sense

Reply:
@emhane I'm not so familiar with the codebase and `EthStream`, but from what I understand we have `EthStream<P2PStream<S>>` and the sink api is just calling the P2PStream impl, so I guess that should work? Maybe I'm missing something here!